woocommerce/woocommerce
# add: new core profiler scaffolding in xstate

### Changes proposed in this Pull Request:



Closes 194-gh-woocommerce/team-ghidorah.

In this PR the rudimentary scaffolding for the new core profiler is introduced. The development work in this feature will also serve as an experimental testing ground for the usage of XState as a workflow framework. 

As this is only a rudimentary scaffold, it should not be visible to end users and the UI work included is minimal. There will be follow up PRs that will work on individual pages as well as other relevant features to make this feature complete.

This new core profiler is locked behind the feature flag `'core-profiler'` and is accessible by directly visiting this URL only: http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fprofiler



### How to test the changes in this Pull Request:

This feature is not required to be tested as the follow up PRs will have more comprehensive testing instructions.

For now, testing should determine that this feature is correctly feature flagged and the existing onboarding wizard still functions.

1. Enable the `'core-profiler'` feature flag using the WC Test Helper in WooCommerce Beta Tester plugin
2. Visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fprofiler to check that it loads
3. Feel free to step through it.
4. If you wish to see the XState visualizer, uncomment the code that invokes `@xstate/inspect`

https://user-images.githubusercontent.com/27843274/231448163-73b997dd-7c14-4abe-9db0-bfa34b574a82.mov





### Reviewer's Focus

Reviewers of this PR should focus on the usability of the XState visualizer, and that the state machine matches up with the expected flows described in the Figma: Y5pUYSJPsGEud1vknUZhi8-fi-1-147&t=jNYlBt8fZ2IyKff1-0

The individual page event handlers should also be inspected to see if the provided event handlers and context information is sufficient to develop the pages. Optionally, these can also be improved as part of the development of the pages.

### Follow up work:

#### Supporting work on the state machine
1. Navigation Component 193-gh-woocommerce/team-ghidorah
2. Geolocation - Either using JS (203-gh-woocommerce/team-ghidorah) or the wccom geolocation API, determine the user's store location. The state of this should be tracked in a parallel state, beside the main workflow. The conditional guard in preBusinessInfo should also be updated to rely on this state.
3. [Extensions Fetching](https://github.com/woocommerce/woocommerce/issues/37992) - This should tie in with the existing @woocommerce/data/onboarding redux code to retrieve the available extensions. The state of this should be tracked in a parallel state, beside the main workflow. The conditional guard in preExtensions should be updated to rely on this state.
4. [URL Parsing](https://github.com/woocommerce/woocommerce/issues/37993). The state machine should be aware of the URL paths and update the paths accordingly when the pages are stepped through. When an individual page URL is navigated to directly, the initialization state should detect this and fire off the appropriate transition to redirect the user. The development of this should also consider the future work of accepting the Jetpack callback
5. Jetpack redirect out and callback. Details TBD
8. [Style the core profiler page](https://github.com/woocommerce/woocommerce/issues/37994) (without any pages rendered, so perhaps a spinner)
9. [Implement model based testing](https://github.com/woocommerce/woocommerce/issues/37995) https://css-tricks.com/model-based-testing-in-react-with-state-machines/
10. [Dev tooling improvements](https://github.com/woocommerce/woocommerce/issues/37996) e.g [eslint-plugin-xstate](https://www.npmjs.com/package/eslint-plugin-xstate), [dev env detection](https://github.com/woocommerce/woocommerce/pull/37628#discussion_r1165103112) and [typegen](https://github.com/woocommerce/woocommerce/pull/37628#discussion_r1165103308)

#### Pages
1. introOptIn  "Welcome to Woo!" Issue: 195-gh-woocommerce/team-ghidorah
2. userProfile "Which one of these best describes you?" Issue: 196-gh-woocommerce/team-ghidorah
3. businessInfo "Tell us a bit about your store" Issue: 197-gh-woocommerce/team-ghidorah
4. businessLocation "Where is your business located?" 198-gh-woocommerce/team-ghidorah
